src/functional/ffront/func_to_foast.py

```diff
@@ -16,23 +16,153 @@
+    try:
+        filename = inspect.getabsfile(func) or MISSING_FILENAME
+        source = textwrap.dedent(inspect.getsource(func))
+        starting_line = (
+            inspect.getsourcelines(func)[1] if not filename.endswith(MISSING_FILENAME) else 1
+        )
+    except OSError as err:
+        if filename.endswith(MISSING_FILENAME):
+            message = "Can not create field operator from a function that is not in a source file!"
+        else:
+            message = f"Can not get source code of passed function ({func})"
+        raise ValueError(message) from err
+
+    return SourceDefinition(source, filename, starting_line)
+
+
+def _make_closure_refs_from_function(func: Callable) -> ClosureRefs:
+    (nonlocals, globals, inspect_builtins, inspect_unbound) = inspect.getclosurevars(  # noqa: A001
+        func
+    )
+    # python builtins returned by getclosurevars() are not ffront.builtins
+    unbound = set(inspect_builtins.keys()) | inspect_unbound
+    builtins = unbound & set(fbuiltins.ALL_BUILTIN_NAMES)
+    unbound -= builtins
+    annotations = typingx.get_type_hints(func)
+
+    return ClosureRefs(nonlocals, globals, annotations, builtins, unbound)
+
+
+def _make_symbol_names_from_source(source: str, filename: str = MISSING_FILENAME) -> SymbolNames:
+    try:
+        mod_st = symtable.symtable(source, filename, "exec")
+    except SyntaxError as err:
+        raise common.GTValueError(
+            f"Unexpected error when parsing provided source code (\n{source}\n)"
+        ) from err
+
+    assert mod_st.get_type() == "module"
+    if len(children := mod_st.get_children()) != 1:
+        raise common.GTValueError(
+            f"Sources with multiple function definitions are not yet supported (\n{source}\n)"
+        )
+
+    func_st = children[0]
+
+    param_names = set()
+    imported_names = set()
+    local_names = set()
+    for name in func_st.get_locals():
+        if (s := func_st.lookup(name)).is_imported():
+            imported_names.add(name)
+        elif s.is_parameter:
+            param_names.add(name)
+        else:
+            local_names.add()
+
+    # symtable returns regular free (or non-local) variables in 'get_frees()' and
+    # the free variables introduced with the 'nonlocal' statement in 'get_nonlocals()'
+    nonlocal_names = set(func_st.get_frees()) | set(func_st.get_nonlocals())
+    global_names = set(func_st.get_globals()) - set(fbuiltins.ALL_BUILTIN_NAMES)
+
+    return SymbolNames(
+        params=param_names,
+        locals=local_names,
+        imported=imported_names,
+        nonlocals=nonlocal_names,
+        globals=global_names,
+    )
+
+
+@dataclass(frozen=True)
+class SourceDefinition:
+    source: str
+    filename: str = MISSING_FILENAME
+    starting_line: int = 1
+
+    def __iter__(self) -> Iterator[tuple[str, ...]]:
+        yield from iter((self.source, self.filename, self.starting_line))
+
+    from_function = staticmethod(_make_source_definition_from_function)
```

Comment:
What's the thought process behind this pattern?

Reply:
The `__iter__` method is added here just to support convenient unpacking of the `SourceDefinition` class. Example:

```python
srd_def = SourceDefinition(source, filename, starting_line)
source, filename, starting_line = src_def
```

Reply:
I was referring more to the

```python
from_function = staticmethod(...)
```

pattern, rather than 

```python
@staticmethod
def from_function(...):
    ...
```

Reply:
Ah, ok, sorry about the confusion. I decided to separate the implementation of the creation method from the class to enhance readability, since methods are not very short and inlining them in the class body polluted the simplicity of the dataclass definition with code that is only related to its creation, not in how the object behaves.

Reply:
That's fine with me. The only thing that might cause some confusion is that we are creating a protected function and then we are "leaking" it into the public API. The uninitiated dev might assume that the signatures there are implementation details.

Reply:
You're right, I made them public to avoid inconsistencies.